### PR TITLE
Completion fix shtab

### LINKED ISSUE #134: tldr: error: argument command: invalid choice
Affects current `master`

```
$ python --version
3.7.6
$ pip install 'git+https://github.com/tldr-pages/tldr-python-client@master#egg=tldr'
$ tldr ls
tldr: error: argument command: invalid choice: 'ls' (choose from [])
```

### PR DESCRIPTION
**Pull request by casperdcl**
- fix #134 invalid choice
- use [shtab](https://github.com/iterative/shtab/) for faster completion

### CODE REVIEW

**Review by sbrl — APPROVED** [review]
Looks ok to me, but I'm not a Python expert.

/cc @MasterOdin for checking, I think this one slipped through the net.

**Comment by MasterOdin** [comment]
Sorry for not reviewing this sooner. I do not have much of an opinion on argcomplete vs shtab, and I've found that both work well on my `oh-my-zsh` installation on my macOS.

However, I am not keen on adding a dependency that has a classifier of `Development Status :: 3 - Alpha` (shtab) over one marked `Development Status :: 5 - Production/Stable` (argcomplete).

**Comment by casperdcl** [comment]
@MasterOdin `shtab` is actually faster than `argcomplete` and has no side effects (it doesn't execute the program every time `<TAB>` is pressed). It also supports all `zsh` features natively in addition to `bash`.

I think `shtab` is now mature enough that it could be classified as stable or at least beta. It's used by https://github.com/iterative/dvc.

Feel free to modify this PR (e.g. merge #149 into #135) if you'd like.

**Comment by casperdcl** [comment]
just tagged `shtab` as `Development Status :: 4 - Beta` and merged in #149 :)

**Comment by casperdcl** [comment]
not sure why the 3.5 build is failing - `shtab`'s `pyproject.toml` clearly states `setuptools>=42` but for some reason https://github.com/tldr-pages/tldr-python-client/pull/135/checks?check_run_id=1618096569 is complaining about `setuptools<30.4`.

**Comment by MasterOdin** [comment]
Testing this out, I get the following error:

```
tldr
_arguments:comparguments:325: doubled rest argument definition: *::args:->args
```

I am on macOS 11, using zsh 5.8 and oh-my-zsh.

**Comment by casperdcl** [comment]
@bl-ue just rebased

**Comment by bl-ue** [comment]
ping @MasterOdin — did you forget about this?

**Comment by MasterOdin** [comment]
Sorry @casperdcl for not following up here. Will run through with a comparison of shtab and argcomplete again soon.

**Comment by casperdcl** [comment]
Sure, also just released a new `shtab` version ([v1.3.10](https://github.com/iterative/shtab/releases/tag/v1.3.10)) so both bash & zsh completion should work perfectly now

**Comment by casperdcl** [comment]
just rebased & added `cmd` completion too (https://github.com/tldr-pages/tldr-python-client/pull/135/commits/c9db2c093853c4f12806ecdf2994c39ee083b8f9)

File: setup.py [hunk 1]
```diff
@@ -26,7 +26,7 @@
             "tldr = tldr:cli"
         ]
     },
-    install_requires=['termcolor', 'colorama', 'argcomplete'],
+    install_requires=['termcolor', 'colorama', 'shtab>=1.3.6'],
```
Comment (casperdcl) [inline setup.py:29]:
```suggestion
    install_requires=['termcolor', 'colorama', 'shtab>=1.3.10'],
```

optional for full bash support

File: tldr.py [hunk 2]
```diff
@@ -434,10 +434,17 @@ def main():
                         help='Override the default language')
 
     parser.add_argument(
-        'command', type=str, nargs='*', help="command to lookup", metavar='command'
-    ).completer = argcomplete.completers.ChoicesCompleter(get_commands())
+        'command', type=str, nargs='*', help="command to lookup", metavar='command',
```
Comment (casperdcl) [inline tldr.py:437]:
```suggestion
        'command', type=str, nargs='*', help="command to lookup", metavar='command'
```

File: tldr.py [hunk 3]
```diff
@@ -434,10 +434,17 @@ def main():
...
+    ).complete = {"bash": "shtab_tldr_cmd_list", "zsh": "shtab_tldr_cmd_list"}
+
+    shtab.add_argument_to(parser, preamble={
+        "bash": '''shtab_tldr_cmd_list(){{
+          compgen -W "$("{py}" -m tldr --list | sed 's/\W/ /g')" -- "$1"
+        }}'''.format(py=sys.executable),
+        "zsh": '''shtab_tldr_cmd_list(){{
```
Comment (casperdcl) [inline tldr.py:444]:
```suggestion
        'bash': '''shtab_tldr_cmd_list(){{
          compgen -W "$("{py}" -m tldr --list | sed 's/\W/ /g')" -- "$1"
        }}'''.format(py=sys.executable),
        'zsh': '''shtab_tldr_cmd_list(){{
```

in case you prefer single quotes

**Review by MasterOdin — APPROVED** [review]

**Comment by MasterOdin** [comment]
Hm, looks like having flake errors. I think to resolve them, will need to double escape the sequence `\\W`? Not sure, will look at this shortly.

**Comment by casperdcl** [comment]
@MasterOdin yes, fixed.
